packages/messaging/test/sw-controller.test.ts

```diff
@@ -638,162 +638,164 @@ describe('Firebase Messaging > *SwController', () => {
       expect(event.notification.close.callCount).to.equal(1);
     });
 
-    it('should open click_action URL for FCM notification (third-party so no window client access)', async () => {
-      const clickAction = '/test-click-action';
-      const waitUntilSpy = sandbox.spy();
-      const event: any = {
-        notification: {
-          data: {
-            FCM_MSG: {
+    ['link', 'click_action'].forEach(key => {
+      describe(key, () => {
+        let link: string;
+        let waitUntilSpy: sinon.SinonSpy;
+        let event: any;
+
+        beforeEach(() => {
+          link = '/test-link';
+          waitUntilSpy = sandbox.spy();
+          if (key === 'link') {
+            // New API
+            event = {
               notification: {
-                click_action: clickAction
-              }
-            }
-          },
-          close: sandbox.spy()
-        },
-        waitUntil: waitUntilSpy,
-        stopImmediatePropagation: sandbox.spy()
-      };
+                data: {
+                  FCM_MSG: {
+                    notification: {},
+                    fcmOptions: {
+                      link: link
+                    }
+                  }
+                },
+                close: sandbox.spy()
+              },
+              waitUntil: waitUntilSpy,
+              stopImmediatePropagation: sandbox.spy()
+            };
+          } else if (key === 'click_action') {
+            // Legacy API
+            event = {
+              notification: {
+                data: {
+                  FCM_MSG: {
+                    notification: {
+                      click_action: link
+                    }
+                  }
+                },
+                close: sandbox.spy()
+              },
+              waitUntil: waitUntilSpy,
+              stopImmediatePropagation: sandbox.spy()
+            };
+          }
+        });
 
-      const openWindowStub = sandbox.stub().callsFake(() => {
-        // Pretend it's a third party URL
-        return null;
-      });
+        it('should open URL (third-party so no window client access)', async () => {
```

Comment:
nit: Alec always suggested the "opens URL ..." form over "should open URL ...". It's shorter, and reads better when you consider that this is a property of the SUT that is either true or not? :stuck_out_tongue: